### ISSUE TITLE
Triggering performUpdate() seems a lot less sensitive than previous version 2.4.0

### ISSUE THREAD

**Issue by cintaccs** [Polymer/lit-element#1064]
With the new updated 2.4.0 version most of the functionality in my bigger app build on lit element 2.3.1 stopped refreshing content. It is a bigger complex application with many levels of lit-element based templates nested. Using Provider and requester models, event propagation etc. as "described" and "recommended" (to state it short - I believe I have a rather standard implementation with lit-element, MDC web components etc.).

First I found that I missed a higher level property type Object - it was enough previously to have this Object as a constructor member and as property on the underlying nested template HTML tag - I added that - got a little further down - but now I have another where refreshing a list element under tabbed content (using MDC-tab lit based component) fails to refresh.

In another case I have a component factory previously returning lit element tagged content which no longer seems to update.

I looked at the change log - but I see Added stuff for property - but not anything indicating that previous properties no longer should trigger refreshing.

For most parts I am not using slots - and I am using shadow DOM on all mentioned above.
It "compiles" with my previous webpack setup - I can "log in" and the application in general "render" but as soon as I trigger events - it is less likely to re-render as with the previous version. I use ES6 js only (no TypeScript, no Babel).

Any idea what could be wrong? What changed? Where I can look for the missing link?

**Comment by justinfagnani** [comment]
Hi @cintaccs sorry for the problems. Can you help out with some more information?

> First I found that I missed a higher level property type Object

What do you mean by this? Object is a built-in, you shouldn't be able to be missing it.

> returning lit element tagged content

What is "tagged content"?

Is there any chance of getting a reduced reproduction?

**Comment by cintaccs** [comment]
"higher level Object":
In my app level component I have a "state" kind of Object (drill) defined on the app constructor. Previously it was not needed to have that "state" member set as a property on the app element. With the new version - that seems to at least make some difference in order to trigger my page stack to load new content. 

Works fine in 2.3.1
````
   static get properties() {
        return {
            mobile: { type: Boolean },
            draweropen: { type: Boolean },
            modal: { type: Boolean },
            dialog: { type: Boolean },
        };
    }

    constructor() {
        super();
        this.drill = null;
        this.draweropen = true;
        this.skipChanges = false;
        this.pendingEvent = null;
...````

drill needs to be on properties in 2.4 although it is never delivered to the app from a higher parent. Like this:
````
   static get properties() {
        return {
            mobile: { type: Boolean },
            draweropen: { type: Boolean },
            modal: { type: Boolean },
            dialog: { type: Boolean },
            drill: { type: Object }
        };
    }

    constructor() {
        super();
        this.drill = null;
        this.draweropen = true;
        this.skipChanges = false;
        this.pendingEvent = null;
...````


In other words - previously the "drill" Object didn't have to be part of properties on my outermost app component - the inner changes to the Object was enough to re-render the factory on the main app element.

**Comment by cintaccs** [comment]
With tagged content I mean - when I have a render function like:

```

    render() {
        const mainStyle = styleMap({
            height: this.asdialog ? 'calc(80vh - ' + (this.noListheader ? 'var(--headerheightbasewithtabs)' : '149px') + ')' : 'calc(100vh - ' + this.height + ')',
            overflow: 'hidden',
            width: 'inherit',
            marginTop: this.sublist ? '0px' : this.height
        });
        console.log('render list');

        return html `
        <dynamicai-header
            .baseModel=${this.drill.baseModel}
            .parentHeader=${!this.sublist}
            .grouped=${this.drill.grouped}
            .tab=${this.drill.tab}
            .loading=${this.loading} 
            .asdialog=${this.asdialog} 
            .draweropen=${this.draweropen} 
            .showback=${(this.pageStack.length > 1) ||  this.groupedBackSteps > 0}
            .search="${this.drill.search}"
            .listheader=${!this.noListheader}
            .subHead=${this.drill.subHead}
            .options=${this.baseNodeType && this.baseNodeType.options}
            .add=${this.baseNodeType && !this.baseNodeType.updatesOnly ? (this.baseNodeType.add || [{ actionIcon: "add", items: [{ viewtype: "add" }] }]) : false}
            .pageName="${this.pageName}"
            .selected=${(this.drill.selected ? Array.from(this.drill.selected) : [])}
            .anyItems=${this.drill.items && this.drill.items.length > 0}>
        </dynamicai-header>
        <main style=${mainStyle}>
            ${this.UI[this.nodetype] ? this.menutype : this.output}
        </main>
        ${!this.dialog? nothing : dialogFactory(this.drill.baseModel, this.dialogEvent)}
        `;
    }
```

this.output is a getter - that return another getter with either list or cards content.

**Comment by cintaccs** [comment]
@justinfagnani  Thanks a lot for looking into this!
I am willing to try to make a reduced reproduction - but at the moment I am not sure which part is not working because it seems as if the element life-cycle is altered in a way that my app logic doesn't see...
Just wanted to give you early feedback - as it in my case at least seems like breaking previously working stuff.

**Comment by cintaccs** [comment]
one thing that I do a lot which might be related to the issue is:
<my-list  .viewtype=${this.viewtype}></..

then some drawer or other parent element will change this.viewtype and I will have something like:
```
updated(changedProperties) {
        let updateIt = false;
        if (this.drill.fromDrawer) {
            this.drill.fromDrawer = false;
            updateIt = true;
        }

        if ((this.nodetype + '/' + this.viewtype + '/' + this.tab + '/' + this.changeStackCount + '/' + this.sideopen) !== this.drillprevious) {
            updateIt = true;
        }

        if (updateIt) {
                this.drillprevious = this.nodetype + '/' + this.viewtype + '/' + this.tab + '/' + this.changeStackCount + '/' + this.sideopen;
                this.refreshItems();
                return this.performUpdate();
           }
```

with the new version 2.4 I see that the component calls the render() - then it will get into the updated function and depending on the state of the variables (this.viewtype e.g.) it will do this.performUpdate() - but it doesn't actually re-render the new content any longer.

**Comment by sorvell** [comment]
Really sorry for the breakage and thanks for your patience working through it. 

Although this was not necessarily super clear, previously to 2.4 we did not intend for users to call `this.performUpdate()` directly as you have done here. Given that, we did make a slight change to `performUpdate` so that it does nothing unless an update has already been requested. As called out in the changelog, we *do* now intend for `performUpdate` to be callable to force any pending update to synchronously "flush."

So, you should be able to change the line that calls `return this.performUpdate()` to just `this.requestUpdate()`. That will schedule an update at the next microtask which should be fine, but if for some reason this must be synchronous, you can call `this.performUpdate()` after calling `this.requestUpdate()`.

I hope that fixes a lot of the issues you're seeing.

**Comment by cintaccs** [comment]
@sorvell That's it!! Perfect! I replaced the performUpdate with requestUpdate in 30+ components (I apparently really like performUpdate... :-0) and apart from 1 location I did not have to add this.performUpdate at all. After a quick round of testing only 1-2 issues - which I think is really nothing in comparison to 30+ changes.
I am now on the latest lit-html and lit-element - thanks thanks and thanks for quick and sharp response!!!
(and btw - nothing to be sorry about - I understand and appreciate what you are doing!!) ... and doing really well!!
